source/vibe/web/internal/rest/common.d

```diff
@@ -0,0 +1,407 @@
+		template Impl(size_t idx) {
+			static if (idx < AllMethods.length) {
+				alias R = ReturnType!(AllMethods[idx]);
+				static if (is(R == interface)) {
+					alias Impl = TypeTuple!(AllMethods[idx], Impl!(idx+1));
+				} else {
+					alias Impl = Impl!(idx+1);
+				}
+			} else alias Impl = TypeTuple!();
+		}
+		alias GetSubInterfaceFunctions = Impl!0;
+	}
+
+	private template GetSubInterfaceTypes() {
+		template Impl(size_t idx) {
+			static if (idx < AllMethods.length) {
+				alias R = ReturnType!(FunctionTypeOf!(AllMethods[idx]));
+				static if (is(R == interface)) {
+					alias Impl = TypeTuple!(R, Impl!(idx+1));
+				} else {
+					alias Impl = Impl!(idx+1);
+				}
+			} else alias Impl = TypeTuple!();
+		}
+		alias GetSubInterfaceTypes = Impl!0;
+	}
+
+	private template GetRouteFunctions() {
+		template Impl(size_t idx) {
+			static if (idx < AllMethods.length) {
+				alias F = AllMethods[idx];
+				static if (!is(ReturnType!(FunctionTypeOf!F) == interface))
+					alias Impl = TypeTuple!(F, Impl!(idx+1));
+				else alias Impl = Impl!(idx+1);
+			} else alias Impl = TypeTuple!();
+		}
+		alias GetRouteFunctions = Impl!0;
+	}
+
+	private template GetAllMethods() {
+		template Impl(size_t idx) {
+			static if (idx < memberNames.length) {
+				enum name = memberNames[idx];
+				// WORKAROUND #1045 / @@BUG14375@@
+				static if (name.length != 0)
+					alias Impl = TypeTuple!(MemberFunctionsTuple!(I, name), Impl!(idx+1));
+				else alias Impl = Impl!(idx+1);
+			} else alias Impl = TypeTuple!();
+		}
+		alias GetAllMethods = Impl!0;
+	}
+}
+
+struct Route {
+	string functionName; // D name of the function
+	HTTPMethod method;
+	string pattern; // relative route path (relative to baseURL)
+	string fullPattern; // absolute version of 'pattern'
+	bool pathHasPlaceholders; // true if path/pattern contains any :placeholers
+	PathPart[] pathParts; // path separated into text and placeholder parts
+	Parameter[] parameters;
+	Parameter[] queryParameters;
+	Parameter[] bodyParameters;
+	Parameter[] headerParameters;
+	Parameter[] attributedParameters;
+	Parameter[] internalParameters;
+}
+
+struct PathPart {
+	/// interpret `text` as a parameter name (including the leading underscore) or as raw text
+	bool isParameter;
+	string text;
+}
+
+struct Parameter {
+	ParameterKind kind;
+	string name;
+	string fieldName;
+	bool isIn, isOut;
+}
+
+struct StaticRoute {
+	string functionName; // D name of the function
+	string rawName; // raw name as returned 
+	bool pathOverride; // @path UDA was used
+	HTTPMethod method;
+	StaticParameter[] parameters;
+}
+
+struct StaticParameter {
+	ParameterKind kind;
+	string name;
+	string fieldName; // only set for parameters where the field name can be statically determined - use Parameter.fieldName in usual cases
+	bool isIn, isOut;
+}
+
+enum ParameterKind {
+	query,       // req.query[]
+	body_,       // JSON body
+	header,      // req.header[]
+	attributed,  // @before
+	internal     // req.params[]
+}
```

Comment:
IIUC, internal is for path attribute, isn't it ? Maybe that'll be a more descriptive name. Also body parameters are not meant to stay JSON forever. On that topic: any way to check the server setting ? It's not a big deal, but if  someone disable `parseQueryString` for example, there's no way to detect it.